zui/src/ui/hooks/useFormData.tsx

```diff
@@ -95,53 +95,60 @@ export const useFormData = (fieldSchema: JSONSchema, path: string[]) => {
   const disabledMask = useMemo(() => parseMaskableField('disabled', fieldSchema, data), [fieldSchema, data])
 
   useEffect(() => {
-    context.setHiddenState(setObjectPath(context.hiddenState, path, hiddenMask || {}))
-    context.setDisabledState(setObjectPath(context.disabledState, path, disabledMask || {}))
-  }, [hiddenMask, disabledMask])
+    formContext.setHiddenState((hiddenState) => setObjectPath(hiddenState, path, hiddenMask || {}))
+    formContext.setDisabledState((disabledState) => setObjectPath(disabledState, path, disabledMask || {}))
+  }, [JSON.stringify({ fieldSchema, data })])
 
   const { disabled, hidden } = useMemo(() => {
-    const hidden = hiddenMask === true || getPathData(context.hiddenState, path)
-    const disabled = disabledMask === true || getPathData(context.disabledState, path)
+    const hidden = hiddenMask === true || getPathData(formContext.hiddenState, path)
+    const disabled = disabledMask === true || getPathData(formContext.disabledState, path)
     return { hidden: hidden === true, disabled: disabled === true }
-  }, [context.hiddenState, context.disabledState, hiddenMask, disabledMask, path])
+  }, [formContext.hiddenState, formContext.disabledState, hiddenMask, disabledMask, path])
 
   const handlePropertyChange = useCallback(
     (path: string[], data: any) => {
-      context.setFormData(setObjectPath(context.formData, path, data))
+      formContext.setFormData((formData) => setObjectPath(formData, path, data))
     },
-    [context.formData],
+    [formContext.setFormData],
   )
-
   const addArrayItem = useCallback(
-    (path: string[], data: any) => {
-      const currentData = getPathData(context.formData, path) || []
-      context.setFormData(
-        setObjectPath(
-          context.formData,
-          path,
-          Array.isArray(currentData) ? [...currentData, data] : [...currentData, data],
-        ),
-      )
+    (path: string[], data: any = undefined) => {
+      const defaultData = getDefaultValues((fieldSchema as ArraySchema).items)
+
+      formContext.setFormData((formData) => {
+        const currentData = getPathData(formData, path) || []
+        if (data === undefined) {
+          data = defaultData
+        }
+        return setObjectPath(formData, path, Array.isArray(currentData) ? [...currentData, data] : [data])
+      })
     },
-    [context.formData],
+    [formContext.setFormData],
   )
 
   const removeArrayItem = useCallback(
     (path: string[], index: number) => {
-      const currentData = getPathData(context.formData, path)
-      currentData.splice(index, 1)
-      context.setFormData(setObjectPath(context.formData, path, currentData))
+      formContext.setFormData((formData) => {
+        const currentData = getPathData(formData, path) || []
+
+        if (!Array.isArray(currentData)) {
+          return formData
+        }
+
+        currentData.splice(index, 1)
+        return setObjectPath(formData, path, currentData)
```

Comment:
:eyes: **nit (non-blocking):** instead of mutating currentData you could filter it instead (same behavior as the add): `currentData.filter((_, i) => i !== index)`